### ISSUE TITLE
Color Plugin: ColorConverter should not convert null values

### ISSUE THREAD

**Issue by rusanov-vladimir** [MvvmCross/MvvmCross#3632]
## 🐛 Bug Report

After upgrading to VS2019 16.3 handled exceptions are causing performance overhead when debugger is attached. When using color plugin binding and `ToNative` value converter I'm getting handled exceptions when color value is null.

For example: 
`local:MvxBind="BackgroundColor NativeColor(Header.BackgroundColor);"`

Exception is thrown at [this](https://github.com/MvvmCross/MvvmCross/blob/8987384b7cc495410bca279ecf4be8c60c0152f0/MvvmCross.Plugins/Color/MvxColorValueConverter.cs#L31) line.
It's handled somewhere, so it's not crashing the application, but when there is a view with a lot of such bindings and values are mostly null - it takes minutes to render that view with debugger attached.

Checking for value not being null before trying to Convert it seems to be the obvious way to fix it.

### Expected behavior
Converter should not throw and handle exception when value is null.

### Reproduction steps
1. Create a binding like:
`local:MvxBind="BackgroundColor NativeColor(Header.BackgroundColor);"`
2. Ensure model property `Header` is null
3. Exception Settings -> Common Languange Runtime Exceptions should be marked

### Configuration

**Version:** 6.4.1

**Platform:** 
- [ ] :iphone: iOS
- [X] :robot: Android
- [ ] :checkered_flag: WPF
- [ ] :earth_americas: UWP
- [ ] :apple: MacOS
- [ ] :tv: tvOS
- [ ] :monkey: Xamarin.Forms

**Comment by Cheesebaron** [comment]
Would you be able to make the PR and fix this? 

I guess it would be nicely fixed by doing something like:

```csharp
if (value is T t)
    return Convert(t, parameter, culture);
return default(System.Drawing.Color);
```

Not sure if `T` needs to be restricted for this to work.

**Comment by rusanov-vladimir** [comment]
I'm not sure that returning default color would make sense in this case. If the color value is null, then I would prefer not to change the color on the widget, rather to set it to default color.
To work around it in my project I've uninstalled Color plugin and created my own target bindings for `BackgroundColor` and `TextColor` like this:

    public class ViewTargetBinding : MvxAndroidTargetBinding<View, Color?>
    {
        public const string BackgroundColor = "BackgroundColor";

        public ViewTargetBinding(View target)
            : base(target)
        {
        }

        protected override void SetValueImpl(View view, Color? value)
        {
            if (value == null)
                return;
            view.SetBackgroundColor(AppUtility.ToAndroidColor(value.Value));
        }
        protected override Color? MakeSafeValue(Color? value)
        {
            return value ?? null;
        }

        public override MvxBindingMode DefaultMode => MvxBindingMode.OneTime;
    }

This way I don't even need a value converter and I can bind directly like this:
`local:MvxBind="TextColor Header.TextColor; BackgroundColor Header.BackgroundColor;"`

I don't feel that I have enough knowledge at this point to make a PR to fix this behavior in the library.

**Comment by pinkysek** [comment]
@rusanov-vladimir I looked at this problem and I can reproduce only if BackgroundColor in your Header object is null. But if whole object Header is null it works well. So I made pull request which fix if value in ColorConverter is null.

**Comment by rusanov-vladimir** [comment]
@pinkysek @martijn00 
I still think that this change suggested by @Cheesebaron isn't good, because it can cause unwanted behaviour when the color value is null.
If you have a control, that has `White` background, and you make a binding that has a null value, I would expect it to leave the background `White`. Current implementation was leaving it `White`, but throwing an exception, which was the main issue.
With the implementation that was merged, the background from `White` will become default: `Black` when the value is null. 
I think that it's even worse then it was before!

**Comment by martijn00** [comment]
Hmm, I see. Can you make a PR to change that?

**Comment by rusanov-vladimir** [comment]
At this moment I don't see a good way to approach it. ColorConverter, when is called, MUST return some value or to throw. When the color to convert is null, then ColorConverter should return the current color of the property it's binded to. I don't think we can get that information inside the converter... Throwing exception that was caught somewhere later isn't that bad (I hope MSFT will fix debugger's performance issues sooner or later), so if there are no other bright ideas it's better to leave it throwing exceptions.

**Comment by pinkysek** [comment]
When I tested implementation of `return default(System.Drawing.Color)` on Android platform in Playground.Android project. Background color has been set as default not `Black`. Not sure how other platforms.

**Comment by pinkysek** [comment]
This is how it [looks](https://ibb.co/n80JR1m).

With:
```
ColorHolder = new ColorHolder()
{
     BackgroundColor = null
};

local:MvxBind="BackgroundColor NativeColor(ColorHolder.BackgroundColor);"

```

**Comment by rusanov-vladimir** [comment]
@pinkysek From what I understand, default for struct is a struct with all fields set to their default values. So for color ARGB will be set to 0. Color with RGB set to 0 is `Black`. You can see it in this [fiddle](https://dotnetfiddle.net/itgGSw). According to [MSDN](https://docs.microsoft.com/en-us/dotnet/api/system.drawing.color.a?view=netframework-4.8#remarks): 

> The alpha component specifies the transparency of the color: 0 is fully transparent, and 255 is fully opaque. 

So returning a default color is returning a fully transparent `Black` color. Could you try to check what happens if you have a custom background (`Green`, for example) and have a binding with null? Will it keep the previous background color?

**Comment by pinkysek** [comment]
I tried set custom color and binding null and you are right it's fully transparent. I think it's not worse than before and problem with throwing exception is solved.

**Comment by rusanov-vladimir** [comment]
I think it's a **BREAKING CHANGE**, to say the least, because it was behaving differently before. The exception is not happening indeed, but now the behavior is not that one I'd be expecting. I don't want to color to be changed in case I'm binding to a null property.
Thanks for taking your time and trying to help with this annoying issue, but for my project this change will just make my life a bit harder...

**Comment by pinkysek** [comment]
Your idea with set current color is on Android platform possible but probably not for other platforms.

We have [HERE](https://github.com/MvvmCross/MvvmCross/blob/11798c97af610624c8054dd2302d5d9cb747f386/MvvmCross.Plugins/Color/Platforms/Android/BindingTargets/MvxViewBackgroundColorBinding.cs#L20) and [HERE](https://github.com/MvvmCross/MvvmCross/blob/11798c97af610624c8054dd2302d5d9cb747f386/MvvmCross.Plugins/Color/Platforms/Android/BindingTargets/MvxTextViewTextColorBinding.cs#L20) information of current color.

**Comment by rusanov-vladimir** [comment]
@pinkysek Yes, we have access to target object indeed, but we don't know which property are we converting now. Imagine a control that has `Background` and `Foreground` properties. When you are doing conversion inside the converter you have your target view, but there is no way to know are you setting value for `Background` or for `Foreground`.

**Comment by pinkysek** [comment]
@martijn00 @rusanov-vladimir @Cheesebaron 
And what about add code below [HERE](https://github.com/MvvmCross/MvvmCross/blob/11798c97af610624c8054dd2302d5d9cb747f386/MvvmCross/Binding/Combiners/MvxValueConverterValueCombiner.cs#L82).

```
if (sourceValue == null)
{
     value = MvxBindingConstant.DoNothing;
     return true;
}
```
It's possible make it only for MvxColorValueConverter or make it little better. But it is my fast idea.
This works how @rusanov-vladimir expected first.

**Comment by pinkysek** [comment]
or maybe better change 

`if (sourceValue == MvxBindingConstant.DoNothing)` 
for 
`if (sourceValue == MvxBindingConstant.DoNothing || sourceValue == null)`

I can look at this later more...

**Comment by rusanov-vladimir** [comment]
@martijn00 @Cheesebaron
Please consider reverting this change until it has a proper fix! 
You are going to introduce **breaking changes**, that are much worse then the benifit of not having a handled exception being thrown!
